### DISCUSSION TITLE
onDataChanged callback

### DISCUSSION THREAD

**Discussion by phatmann** [TanStack/query#2911] [Ideas] [answered]
Calling `QueryClient.setQueryData` method triggers the `onSuccess` callback. For many users, including me, this is unexpected behavior, and can be the source of recursion issues, since the `onSuccess` handler might make calls to `setQueryData`.

I propose we create a new `onDataChanged` callback. This would be trigged when cache data is changed via `setQueryData` or via a fetch.  The `onSuccess` callback would **only** be called after a successful fetch.

For completeness, the same `onDataChanged` callback would be added to QueryCache.

**Comment by TkDodo** [comment, marked as answer]
I agree with this - let’s include that in the v4 plan. 

Follow up question: since this is a callback on observer level, would this callback also be called when data changes for the observer because it was read from the cache?
For example, many people use onSuccess to put the data into local state or react context. While this is not really necessary, you can still do it. However, if you combine this approach with a high staleTime, if the query key changes, you’ll get no background refetch - just the data from the cache. So onSuccess is also _not_ called, which puts you out of sync.
An onDataChanged callback could solve that, conceptually becoming an effect with `data` as dependency. What do you think?

**Reply by phatmann** [reply]
Can you give a more concrete use case? I would like to make sure I understand the question fully.

**Reply by TkDodo** [reply]
```
useQuery(
  ['key', id],
  queryFn,
  {
    staleTime: 1000 * 60 * 20,
    onSuccess: data => dispatch({ type: 'DATA_RECEIVED', payload: { data } })
  }
)
```

people do this a lot, and it _sometimes_ makes sense: you call the query at the top level, then distribute the data with context down to your component tree.

Problem here is:

- id = 1, you fetch, onSuccess is called, dispatch
- id = 2, you fetch, onSuccess is called, dispatch
- id =1, no fetch, no onSuccess because staleTime is set and data is read from cache only

Now this wouldn't happen with `staleTime: 0` because the key change triggers a background refetch, but with a staleTime, it does occur.

The better fix at the moment, apart from _not_ syncing state, is:

```
const { data } = useQuery(
  ['key', id],
  queryFn,
  {
    staleTime: 1000 * 60 * 20,
  }
)

React.useEffect(() => {
  if (data) {
    dispatch({ type: 'DATA_RECEIVED', payload: { data } })
  }
}, [data, dispatch])
```

which is more code but works because the effect will be triggered whenever data "changes". The `onDataChanged` callback _could_ also handle that case (not sure if it should / if you agree with this). What do you think?

**Reply by matthewdavi** [reply]
@TkDodo Thanks so much for your work on react-query and your great blog posts. 

I would like to see something like the onDataChanged implemented. There are certain side effects that I would like to have synchronized with my react query state (auto populating form fields with responses from queries, displaying toast notifications if a user selects an invalid set of options, etc, etc). Currently the best solution for this is either to set {staleTime: 0} (and trigger an additional network request), or put it in a useEffect (which I find verbose and prone to bugs). Is this something that could be added in a near-future version of react-query? I'd be happy to try to implement this feature if you're open to accepting contributions. 

Thanks.

**Reply by TkDodo** [reply]
yeah I think this callback might be a good idea, given that the core is framework agnostic now and that `useEffect` just exists in react. If you want to contribute an `onDataChanged` callback, please go ahead and we can discuss details in the PR.

FYI, I think it should run after `select`, so it would receives the transformed data from `select`. This is also how `onSuccess` works.
